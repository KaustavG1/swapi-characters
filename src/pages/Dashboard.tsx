import { useEffect, useState } from "react";
import { AxiosError } from "axios";
import ErrorMessage from "../components/common/ErrorMessage/ErrorMessage";
import Header from "../components/Header/Header";
import ListContainer from "../components/ListContainer/ListContainer";
import Loader from "../components/common/Loader/Loader";
import Pagination from "../components/Pagination/Pagination";
import useFetch from "../hooks/useFetch";
import usePaginatedData from "../hooks/usePaginatedData";
import fetchDataFromUrl from "../utils/fetchData";
import { baseUri, people } from "../constants/constants";
import { PaginationDirection } from "../enums/PaginationDirection";
import { Character } from "../models/Character";
import { CharacterDetails } from "../models/CharacterDetails";
import { PlanetDetails } from "../models/PlanetDetails";

function Dashboard() {
  const [isPlanetLoading, setPlanetLoading] = useState(false);
  const [planetError, setPlanetError] = useState<AxiosError | null>(null);
  const [collatedData, setCollatedData] = useState<Character[]>([]);
  const [searchTerm, setSearchTerm] = useState<string>("");
  const uri = `${baseUri}/${people}`;
  const {
    isLoading,
    data: peopleData,
    error,
  } = useFetch<CharacterDetails[]>(uri);
  const { paginatedData, previous, next, setData, page, totalPages } =
    usePaginatedData(peopleData);

  useEffect(() => {
    if (Array.isArray(peopleData)) {
      setData(peopleData?.slice(0, 14));
    }
  }, [peopleData]);

  const handlePagination = (dir: string) => {
    if (dir === PaginationDirection.prev) {
      previous();
    } else {
      next();
    }
  };

  useEffect(() => {
    if (Array.isArray(paginatedData) && paginatedData?.length > 0) {
      const fetchPlanetData = async () => {
        const reqdDetails: any = [];
        for (const detail of paginatedData) {
          setPlanetLoading(true);
          const { data, error } = await fetchDataFromUrl<PlanetDetails>(
            detail?.homeworld
          );
          setPlanetLoading(false);
          setPlanetError(error as AxiosError);

          reqdDetails.push({
            name: detail?.name,
            uid: detail?.url,
            url: detail?.url,
            hair_color: detail?.hair_color,
            eye_color: detail?.eye_color,
            gender: detail?.gender,
            homeworld: data?.name,
          });
        }

        setCollatedData(reqdDetails);
      };

      fetchPlanetData();
    }
  }, [paginatedData]);

  const getCollatedData = () => {
    return collatedData?.filter((e) =>
      e.name.toLowerCase().includes(searchTerm.toLowerCase())
    );
  };

  if (error || planetError) {
    return <ErrorMessage />;
  }

  return (
    <>
      {isLoading || isPlanetLoading ? (
        <Loader />
      ) : (
        <>
          <Header searchTerm={searchTerm} setSearchTerm={setSearchTerm} />
          <ListContainer data={getCollatedData()} />
          <Pagination
            disablePrev={page < 1}
            disableNext={page === totalPages - 1}
            onPreviousClick={handlePagination}
            onNextClick={handlePagination}
          />
        </>
      )}
    </>
  );
}

export default Dashboard;
