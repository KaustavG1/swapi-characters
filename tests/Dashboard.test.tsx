import axios from "axios";
import { it, expect, vi, MockedFunction } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import Dashboard from "../src/pages/Dashboard";
import { mockPaginationInput } from "./__mocks__/mockPaginationInput";

vi.mock("axios");
(axios.get as MockedFunction<typeof axios>).mockResolvedValue({
  data: mockPaginationInput,
});

afterEach(vi.clearAllMocks);

describe("Dashboard component", () => {
  it("renders 'Dashboard' component success case", async () => {
    const { container } = render(
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>
    );
    const loader = screen.getByText(/Loading .../i);
    expect(loader).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(screen.getAllByText(/Luke Skywalker/i)[0]).toBeInTheDocument();
    });
  });

  it("renders 'Dashboard' component failure case", async () => {
    (axios.get as MockedFunction<typeof axios>).mockRejectedValue(
      new Error("API Error")
    );
    const { container } = render(
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>
    );
    const loader = screen.getByText(/Loading .../i);
    expect(loader).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(
        screen.getAllByText(/An error has occured. Please try again later./i)[0]
      ).toBeInTheDocument();
    });
  });
});
