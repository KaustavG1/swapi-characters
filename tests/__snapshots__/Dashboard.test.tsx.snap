// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterView component > renders 'CharacterView' component failure case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`CharacterView component > renders 'CharacterView' component success case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`Dashboard component > renders 'Dashboard' component failure case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`Dashboard component > renders 'Dashboard' component success case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;
